src/Perpetuum/Zones/SessionlessPlayerTimeout.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Perpetuum.Log;
+using Perpetuum.Players;
+using Perpetuum.Timers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Perpetuum.Zones
+{
+    public class SessionlessPlayerTimeout
+    {
+        private static readonly TimeSpan MAX_ORPHAN_TIME = TimeSpan.FromMinutes(6);
+        private static readonly TimeSpan UPDATE_RATE = TimeSpan.FromMinutes(2);
+        private readonly IZone _zone;
+        private readonly IDictionary<long, PlayerTimeout> _orphanPlayers;
+
+        public SessionlessPlayerTimeout(IZone zone)
+        {
+            _zone = zone;
+            _orphanPlayers = new Dictionary<long, PlayerTimeout>();
+        }
+
+        private TimeSpan _elapsed = TimeSpan.Zero;
+        public void Update(TimeSpan time)
+        {
+            _elapsed += time;
+            if (_elapsed < UPDATE_RATE)
+                return;
+
+            DoUpdate();
+            _elapsed = TimeSpan.Zero;
+        }
+
+        private void DoUpdate()
+        {
+            Logger.DebugInfo($"Zone:{_zone.Id} Total players: {_zone.Players.Count()} Total orphan players: {_zone.Players.Count(p => p.Session == ZoneSession.None)}");
+
+            // Players on the zone without a session - this is ok unless the session never connects!
+            var playersWithoutSessions = _zone.Players.Where(p => p.Session == ZoneSession.None)
+                .Where(p => !_orphanPlayers.Keys.Contains(p.Eid));
+
+            foreach (var p in playersWithoutSessions)
+            {
+                _orphanPlayers.Add(p.Eid, new PlayerTimeout(p));
+            }
+
+            // Players that are no longer orphaned
+            _orphanPlayers.RemoveRange(_orphanPlayers.Where(o => o.Value.Player.Session != ZoneSession.None).GetKeys().ToArray());
+
+            // Players with Expired timers; Zone assumes the Player has failed to connect and removes them
+            var toRemove = _orphanPlayers.Where(o => o.Value.Expired).ToArray();
+
+            foreach (var o in toRemove)
+            {
+                o.Value.Player.RemoveFromZone();
+                _orphanPlayers.Remove(o.Key);
+                Logger.DebugInfo($"A player w/o session REMOVED: {o.Value.Player}");
+            }
+        }
+
+        private class PlayerTimeout
+        {
+            public Player Player { get; private set; }
+            public bool Expired { get { return _time.Expired; } }
+            private readonly TimeKeeper _time;
+
+            public PlayerTimeout(Player p)
+            {
+                _time = new TimeKeeper(MAX_ORPHAN_TIME);
+                Player = p;
+            }
+        }
+
+    }
+}
```

Comment:
First thing first, what you've got here is fine and normally I'd accept it, but the way it is done might want to be "updated" to modern engineering (or at least can be used as a learning experience 👍).

The only difference in particular would be the "exclusive where" clause you have on line 39+40.
What would be nice to see is an `IEqualityComparer<Player>` and for your private class `PlayerTimeout` to be an extension of `Player` itself.

So line 61 would become `private class PlayerTimeout : Player` and would carry on as normal.
You would create a class for the comaprer:
```
public class PlayerComparer : IEqualityComparer<Player>
{
  public bool Equals(Player left, Player right)
  {
    if (Object.ReferenceEquals(left, right)) return true;

    if (Object.ReferenceEquals(left, null) || Object.ReferenceEquals(right, null)) return false;

    return left.Eid == right.Eid;
  }

  public int GetHashCode(Player player)
  {
    if (Object.ReferenceEquals(player, null)) return 0;
    return player.Eid.GetHashCode(); // This will suffice as we're comparing the player references only in this comparer
  }
}
```

So, what this now allows you to do is change your Dictionary to a List (you no longer need the keep access to the keys):
```
var playersWithoutSessions = _zone.Players.Where(p => p.Session == ZoneSession.None).Except(_orphanPlayers, new PlayerComparer());
```

You would also then need to change your List references and make a cast when you're adding it to the list so that you can add the timeout information. But the `PlayerComparer` should come in useful in other scenarios as well

Reply:
As an addendum, you can *also* use the `IEquatable<T>` interface on the `private class PlayerTimeout` class to achieve the same result (you could for example add `IEquatable<Player>` *and* `IEquatable<PlayerTimeout>`). The difference here is that this would make the interface only available via the PlayerTimeout class so you would need to invert clauses (effectively do a Join.Except) so it makes the code a little more complicated than the solution above.

Reply:
I'm a bit wary about extending `Player` because it is sealed, and that seems like an intentional decision with some purpose.  Do you have maybe some insight on what `sealed` is guaranteeing and why something in the codebase might depend on this?